File: contrib/resolver_libs/playdar_resolver.rb
#!/usr/bin/env ruby
# Base resolver class for Ruby
# Subclass PlaydarResolver and override #results and #resolver_settings, 
# we'll take care of the rest (even running it!)

require 'rubygems'
require 'json'

class Object
  def to_json
    JSON.dump(self)
  end
end

class String
    SoundexChars = 'BPFVCSKGJQXZDTLMNR'
    SoundexNums  = '111122222222334556'
    SoundexCharsEx = '^' + SoundexChars
    SoundexCharsDel = '^A-Z'

    # desc: http://en.wikipedia.org/wiki/Soundex
    def soundex(census = true)
        str = upcase.delete(SoundexCharsDel).squeeze

        str[0 .. 0] + str[1 .. -1].
            delete(SoundexCharsEx).
            tr(SoundexChars, SoundexNums)[0 .. (census ? 2 : -1)].
            ljust(3, '0') rescue ''
    end

    def sounds_like(other)
        soundex == other.soundex
    end
end

def print_result(object)
  json = object.to_json
  debug "Returning"
  debug json.inspect
  STDOUT.write [json.size].pack('N')
  STDOUT.write json
  STDOUT.flush
end

def debug(message)
  #STDERR.puts message
end

class PlaydarResolver
  # Override this
  def results(query)
    [{
      :artist => "Queen", 
      :track => "Crazy Little Thing Called Love",
      :size   => 4971780,
      :bitrate => 160,
      :duration => 248,
      :url => "http://127.0.0.1:8081/crazy.mp3",
      :score => 0.99
    }]
  end
  
  # Override this too
  def resolver_settings
    {}
  end
  
  def resolve(query)
    {
      :_msgtype => 'results',
      :qid => query['qid'],
      :results => results(query)
    }
  end
  
  def settings
    {
      :_msgtype => "settings",
      :name => "Generic Ruby Resolver",
      :targettime => 5000,
      :weight => 100
    }.merge(resolver_settings)
  end
  
  def start
    print_result(settings)
    
    catch :done do
      loop while true
    end
  end
  
  def self.start!
    resolver = self.new
    resolver.start
  end
  
  def self.inherited(subclass)
    at_exit{ subclass.start! } unless __FILE__ == $0
  end
  
  def loop
    throw :done unless len_bytes = STDIN.read(4)
    debug "len_bytes #{len_bytes.inspect}"
    throw :done unless length = len_bytes.unpack('N')[0]
    debug "length #{length.inspect}"
    throw :done if length > 4096 || length < 0

    if length > 0
      msg = STDIN.read(length)
      debug "msg #{msg.inspect}"
      begin
        request = JSON.parse(msg)
        result = resolve(request)
        print_result(result)
      rescue Exception => e
        debug "#{e}\n#{e.backtrace.join("\n")}"
      end
    end
  end
end

File: contrib/resolver_libs/playdar_resolver.py
#!/usr/bin/python
"""
Base class for Python resolvers.
"""
import sys, traceback
from httplib import HTTP
# easy_install simplejson
import simplejson as json
from urlparse import urlparse
import urllib, urllib2
import socket
from xml.dom import minidom
from struct import unpack, pack

def soundex(name, len=4):
    """ soundex module conforming to Knuth's algorithm
        implementation 2000-12-24 by Gregory Jorgensen
        public domain
    """

    # digits holds the soundex values for the alphabet
    digits = '01230120022455012623010202'
    sndx = ''
    fc = ''

    # translate alpha chars in name to soundex digits
    for c in name.upper():
        if c.isalpha():
            if not fc: fc = c   # remember first letter
            d = digits[ord(c)-ord('A')]
            # duplicate consecutive soundex digits are skipped
            if not sndx or (d != sndx[-1]):
                sndx += d

    # replace first digit with first alpha character
    sndx = fc + sndx[1:]

    # remove all 0s from the soundex code
    sndx = sndx.replace('0','')

    # return soundex code padded to len characters
    return (sndx + (len * '0'))[:len]

def print_result(o):
	s = json.dumps(o)
	debug("Writing: " + s)
	sys.stdout.write(pack('!L', len(s)))
	sys.stdout.write(s)
	sys.stdout.flush()

def debug(message):
	pass
	#sys.stderr.write(str(message) + "\n")
	#sys.stderr.flush()

class PlaydarResolver:
	def results(self, query):
		"""Method to return results, override this."""
		return []
	
	def resolver_settings(self):
		"""Method to set custom settings, override this."""
		return {}
		
	def settings(self):
		"""Generic settings to be overridden by resolver_settings."""
		debug("settings")
		settings =  {
			'_msgtype':'settings', 
			'name':'Generic Python Resolver', 
			'targettime':5000, 
			'weight':50
		}
		settings.update(self.resolver_settings())
		debug(settings)
		return settings
		
	def resolve(self, query):
		"""Resolve a query, passing it to the subclass's results function."""
		return {'_msgtype':'results', 'qid': query['qid'], 'results':self.results(query)}
	
	@classmethod
	def start_static(cls):
		"""Start from class"""
		return cls().start()
		
	def start(self):
		"""Starts the resolver"""
		print_result(self.settings())
		
		while 1:
			length = sys.stdin.read(4)

			if not length:
					break;

			length = unpack('!L', length)[0]
			if not length:
					break
			# something probably went wrong, most likely we're out of sync and are 
			# reading the 4 bytes length header in the middle of a json string. We can't
			# recover. Bail.
			if length > 4096 or length < 0:
					break
			if length > 0:
					msg = sys.stdin.read(length)
					try:
							request = json.loads(msg)
							# print request
							results = self.resolve(request)
							print_result(results)
					except:
							traceback.print_exc(file=sys.stderr)
							# safe to continue, skipping this msg, because at least
							# we consumed enough input so next iteration hits size header.
							pass